Increment version to 3.0.0-SNAPSHOT

naveentatikonda opened the pull request:
### Description
This PR includes the following changes:

- Increment version on main to 3.0.0-SNAPSHOT
- Nomenclature changes from Whitelist to Allowlist
- Update Lucene 92 package
- Add new KNN Codec KNN940Codec based on Lucene 94 

Nomenclature changes from Whitelist to Allowlist and increment version on main to 3.0.0-SNAPSHOT

 
### Check List
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

navneet1v commented on src/main/java/org/opensearch/knn/index/codec/KNN940Codec/KNN940PerFieldKnnVectorsFormat.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public KnnVectorsFormat getKnnVectorsFormatForField(final String field) {
+        if (isNotKnnVectorFieldType(field)) {
```
isn't this already implemented in older codecs? Can't we use them

martin-gaievski replied:
We may need a refactoring project around codec classes. I think it's better to do it in a separate PR, this is somewhat out of scope for the version bump PR.

naveentatikonda replied:
Yes, agree with Martin. It forced me to add these changes as well to make the build succeed. Else would have added them in a separate PR

martin-gaievski commented on .github/workflows/backwards_compatibility_tests_workflow.yml:
```diff
@@ -46,8 +46,8 @@ jobs:
         java: [ 11, 17 ]
-        bwc_version: [ "1.3.2", "2.0.0", "2.1.0", "2.2.0" ]
-        opensearch_version: [ "2.3.0-SNAPSHOT" ]
+        bwc_version: [ "2.4.0-SNAPSHOT" ]
```
why "2.0.0", "2.1.0", "2.2.0" got removed?

naveentatikonda replied:
For rolling upgrade tests we need to upgrade from most recent version of previous major (2.4.0) to next major version (3.0.0)

martin-gaievski commented on .github/workflows/backwards_compatibility_tests_workflow.yml:
```diff
@@ -14,8 +14,8 @@ jobs:
         java: [ 11, 17 ]
-        bwc_version : [ "1.1.0", "1.2.4", "1.3.2", "2.0.0", "2.1.0", "2.2.0" ]
-        opensearch_version : [ "2.3.0-SNAPSHOT" ]
+        bwc_version : [ "2.0.0", "2.1.0", "2.2.0", "2.3.0", "2.4.0-SNAPSHOT" ]
```
do we need to check against the 2.4 snapshot?

naveentatikonda replied:
Yes. Pls find the below link for explanation
https://github.com/opensearch-project/OpenSearch/issues/3615#issuecomment-1247544266

naveentatikonda commented on src/test/java/org/opensearch/knn/index/codec/KNN940Codec/KNN940CodecTests.java:
```diff
@@ -0,0 +1,40 @@
+
+        Function<PerFieldKnnVectorsFormat, Codec> knnCodecProvider = (knnVectorFormat) -> KNN940Codec.builder()
+            .delegate(createKNN94DefaultDelegate())
+            .knnVectorsFormat((KNN940PerFieldKnnVectorsFormat) knnVectorFormat)
```
ack

naveentatikonda commented:
The BWC Tests from 2.4.0 to 3.0.0 are failing because the artifact URL for unreleased version (2.4.0 (https://ci.opensearch.org/ci/dbc/distribution-build-opensearch/2.4.0/latest/linux/x64/tar/dist/opensearch/opensearch-2.4.0-linux-x64.tar.gz)) is different from the released versions(ex - 2.3.0 (https://artifacts.opensearch.org/releases/bundle/opensearch/2.3.0/opensearch-2.3.0-linux-x64.tar.gz)). I have updated the artifact url on my local and I have added k-NN plugin in the [2.4.0 manifest](https://github.com/opensearch-project/opensearch-build/pull/2653). I'm waiting on the build team to release a new artifact by including k-NN plugin to test the BWC tests. 

Building the new distribution is failing due to some issue with Performance-Analyzer plugin. They are still working on it.

martin-gaievski commented on qa/build.gradle:
```diff
@@ -39,8 +39,15 @@ task pullBwcPlugin {
     doLast {
+        ext{
+            if (knn_bwc_version == "2.4.0") {
+                srcUrl = "https://ci.opensearch.org/ci/dbc/distribution-build-opensearch/${knn_bwc_version}/latest/linux/x64/tar/dist/opensearch/opensearch-${knn_bwc_version}-linux-x64.tar.gz"
```
When we can rollback this logic to a uniform url for all versions? Do you know if build team will move the artifact once 2.4.0 is released, or this url will still work?

naveentatikonda replied:
I think it will still work because when I have tried to download the 2.3.0 version using same url, it worked https://ci.opensearch.org/ci/dbc/distribution-build-opensearch/2.3.0/latest/linux/x64/tar/dist/opensearch/opensearch-2.3.0-linux-x64.tar.gz. But, IMO we should always test against the released version. So, we can comment out this piece of code after 2.4.0 is released

martin-gaievski replied:
Right, that's my concern that we use the actual released artifact and not the latest, as build will work it's on our good intensions to change that URL. We need some sort of action item for ourselves to make sure that change is not abandoned. Could you please create an issue against knn 2.4 to revert url related logic once core 2.4 got finalized for release?

naveentatikonda replied:
Makes sense. Will create an issue pointing to this piece of code which needs to be removed in the future after this PR is merged

navneet1v replied:
I think this will change once we have released version of 2.4 . This change cannot be reverted till we have a maven release for 2.4 or any other version. I feel this change will be there forever in the 2.4 branch, as we should not make the change in 2.4 branch once it is released

martin-gaievski replied:
We may need to revert this in "main", for 2.4 branch - maybe too, it depends on when core team release the final artifact for 2.4 so the url looks similar to other releases.

naveentatikonda replied:
We don't need to make this change in 2.4 branch because we are using this logic and fetching latest artifacts for bwc versions. In 2.4 branch we will only use released versions to upgrade to 2.4.0 version. So, it is not a problem even if the code is not reverted.

heemin32 commented on src/test/java/org/opensearch/knn/index/codec/KNNFormatFactoryTests.java:
```diff
@@ -36,4 +36,16 @@ public void testKNN92Format() {
+        MapperService mapperService = mock(MapperService.class);
+        final Codec lucene94CodecDelegate = KNNCodecFactory.CodecDelegateFactory.createKNN94DefaultDelegate();
+        KNNCodecFactory knnCodecFactory = new KNNCodecFactory(mapperService);
+        final Codec knnCodec = knnCodecFactory.createKNNCodec(lucene94CodecDelegate);
```
Why only Codec is final?

naveentatikonda replied:
updated

martin-gaievski replied:
nit: we can make mapperService and knnCodecFactory final as well, but I guess that's ok, we can do it later

naveentatikonda replied:
knnCodecFactory is a constructor call and mapperService is a mock object. Do we need to make it final ?

martin-gaievski replied:
I think it's a preference, make everything final unless you need to mutate that variable, so you'll get the compilation time error. Those two variable can be final, i will not cause issues.

heemin32 replied:
My main concern here is inconsistency. I cannot find a reason of why mappserService is not final but Codec is final. I would rather remove final from Codec as well unless we have strong reasoning about it. If you see other test cases, we don't define Codec as final.

naveentatikonda replied:
Agree. To make it consistent, I removed final everywhere. There is no strong reason to have it.

martin-gaievski reviewed (approved):
Changes look good

heemin32 commented on src/main/java/org/opensearch/knn/index/codec/KNN940Codec/KNN940PerFieldKnnVectorsFormat.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    public KnnVectorsFormat getKnnVectorsFormatForField(final String field) {
+        if (isNotKnnVectorFieldType(field)) {
+            log.debug(
```
You don't need String.format I think?
```
log.debug("Initialize KNN vector format for field [{}] with default params [max_connections] = \"{}\" and [beam_width] = \"{}\"",
    field,
    Lucene94HnswVectorsFormat.DEFAULT_MAX_CONN,
    Lucene94HnswVectorsFormat.DEFAULT_BEAM_WIDTH
);
```

naveentatikonda replied:
Yes, we can do it the way you suggested without using String.format. But, can you tell me what are the pros and cons of using this over String.format?

heemin32 replied:
pros: less verbose. cons: none that I can think of.

naveentatikonda replied:
Thanks, updated

heemin32 commented on src/main/java/org/opensearch/knn/index/codec/KNN940Codec/KNN940PerFieldKnnVectorsFormat.java:
```diff
@@ -0,0 +1,78 @@
+        var params = type.getKnnMethodContext().getMethodComponent().getParameters();
+        int maxConnections = getMaxConnections(params);
+        int beamWidth = getBeamWidth(params);
+        log.debug(
```
Same here.

naveentatikonda replied:
updated

heemin32 commented on src/main/java/org/opensearch/knn/index/codec/KNN940Codec/KNN940PerFieldKnnVectorsFormat.java:
```diff
@@ -0,0 +1,78 @@
+            return new Lucene94HnswVectorsFormat();
+        }
+        var type = (KNNVectorFieldMapper.KNNVectorFieldType) mapperService.orElseThrow(
+            () -> new IllegalStateException(
```
Don't we have any custom exception?

heemin32 commented on src/test/java/org/opensearch/knn/index/codec/KNNCodecFactoryTests.java:
```diff
@@ -28,12 +29,18 @@ public void testKNN92DefaultDelegate() {
     public void testKNNDefaultCodec() {
         MapperService mapperService = mock(MapperService.class);
         KNNCodecFactory knnCodecFactory = new KNNCodecFactory(mapperService);
-        Codec knnCodec = knnCodecFactory.createKNNCodec(KNNCodecFactory.CodecDelegateFactory.createKNN92DefaultDelegate());
```
Codec is not final here.
